src/app/data-planes/sources.ts

```diff
@@ -120,28 +76,24 @@ export const sources = (api: KumaApi) => {
       })
     },
 
-    '/meshes/:mesh/dataplanes/for/:service/of/:type': async (params: CollectionParams & ServiceParams & PaginationParams & DataplaneTypeParams, source: Closeable) => {
-      source.close()
-
-      const { mesh, size, type } = params
+    '/meshes/:mesh/dataplanes/for/:service/of/:type': async (params) => {
+      const { mesh, size } = params
       const offset = size * (params.page - 1)
 
-      // here 'all' means both proxies/sidecars and gateways this currently fits
-      // our usecases but we should probably include `gateway | sidecar` or
-      // similar
       const filterParams = Object.fromEntries(normalizeFilterFields(JSON.parse(params.search || '[]')))
+
       if (typeof filterParams.tag === 'undefined') {
         filterParams.tag = []
       }
       filterParams.tag = filterParams.tag.filter((item) => !item.startsWith('kuma.io/service:'))
       filterParams.tag.push(`kuma.io/service:${params.service}`)
 
-      const gatewayParams: DataPlaneOverviewParameters = {}
-      if (type === 'standard') {
-        gatewayParams.gateway = 'false'
-      } else if (type !== 'all') {
-        gatewayParams.gateway = type
-      }
+      // we use `all` in code to mean "don't specify ?gateway=0 at all" i.e.
+      // both gateway and sidecars
+      const type = params.type === 'standard' ? 'false' : params.type
+      const gatewayParams = includes(['delegated', 'builtin', 'false'] as const, type)
+        ? { gateway: type }
+        : {}
```

Comment:
**Question**: Why does the implementation of this change? Is this a rebase mistake?

Reply:
Its not a rebase mistake

The reason behind this is: this is what this PR did previous to the change over in the red half of the diff (remember this was PRed way before now). I updated the code to take into account the changes that were made since September when I originally PRed this.

I did spend a tiny bit more time to think about it during rebase to try to make sure it was still correct, which I think it is? No tests blew up at least, and if I change `standard` to `standrd` for example, the tests do blow up.

Here's my thinking:

---

- `?gateway` can only be `delegated`, `builtin`, `false` (or `true` which I don't think we need)
- `standard` means `false` (i.e. don't give me gateways only 'sidecars')
- Anything else (including `all` which is the string we use to say don't add the query param), then don't add `?gateway` at all.

---

Does that sound right to you? Do you think this is correct?

Reply:
I’m following, yes. So in terms of behavior it’s the same as the current implementation only now `gatewayParams` is no longer type-checked against `DataPlaneOverviewParameters`.

Can you please add this again in both places?

Reply:
Here's the type check:

https://github.com/kumahq/kuma-gui/blob/6b453886e8eec507c171134429f5654b01b6865d/src/services/kuma-api/KumaApi.ts#L148

Which checks `params` against `DataPlaneOverviewParameters`

DataSource URLs are just like browser URLs, you can narrow the type check to only certain strings all you like, it doesn't prevent anyone from typing a different thing in the URL apart from the 3 or 4 options you expect. We have this issue URLs in quite a few places still but I'm gradually reducing this issue as I go.

For example with this PR, the following URL no longer sends `?gateway=stadard` to the backend

`/gui/meshes/default/data-planes?page=1&size=50&dataplaneType=stadard`

This is because we do a runtime check on the value first, if its none of the values we expect we just don't pass the `?gateway`. Well, thats the thinking behind it.

Does that sounds right to you? Do you think this is correct?

Reply:
That’s right, but this type check is still valuable here because it prevents us from making mistakes like trying to assign `{ gateways: '...' }` to `gatewayParams`. Without the type check, this would just work.

Reply:
Ok great thanks. So wait, where do you want me to add the additional type check?

Reply:
Right, but `DataPlaneOverviewParameters` includes `true`

> `?gateway` can only be `delegated`, `builtin`, `false` (or `true` which I don't think we need)

> `true` which I don't think we need

We don't ever want to pass `?gateway=true` as far as I'm aware?

So I think this is probably more correct?

![Screenshot 2023-11-24 at 11 16 53](https://github.com/kumahq/kuma-gui/assets/554604/91faeea4-ad84-4bbc-be6d-f62850d6d31c)

If you add something to the `includes` that `DataPlaneOverviewParameters` doesn't have it will be type checked by the code I linked to above I think?

Plus as far as I'm concerned its an awful lot clearer and less error prone (see above), obvs my own opinion on the clarity.

I could also just inline the includes if you'd prefer, something like:

```js
      return api.getAllDataplaneOverviewsFromMesh({ mesh }, {
        ...filterParams,
        ...includes(['delegated', 'builtin', 'false'] as const, type) ? { gateway: type } : {},
        offset,
        size,
      })
```

Let me know, maybe I'm missing something else here?

Reply:
Well, alright, let’s leave it as in the PR I guess.

We could change `DataPlaneOverviewParameters` to not include `'true'` if we don’t ever care about it and still get the type checking benefits within the source definition.

The type check will only work in some cases. It won’t work, for example, when spreading the includes result into the params object:

```ts
type Params = {
  field?: string
}

function fn(obj: Params) {}

// Case 1
fn({
  field: '',
  unexpectedField: '', // <-- ✅ Error
})

// Case 2
fn({
  ...{
    field: '',
    unexpectedField: '', // <-- ❌ No error
  },
})

// Case 3
const b: Params = {
  field: '',
  unexpectedField: '', // <-- ✅ Error
}

// Case 4
const c = {
  field: '',
  unexpectedField: '', // <-- ❌ No error
}
fn(c)
```

Reply:
> Well, alright, let’s leave it as in the PR I guess.

Resolving this one.

Reply:
P.S. Let me know on the rest.